File: frontend/src/services/publicOpeningService.ts
import type { OpeningDetail, OpeningListItem } from '../types/opening';
import type { PageResponse } from '../types/common';
import { api } from './api';

export const publicOpeningService = {
  async getPublicOpenings(
    page = 0,
    size = 20,
    sort = 'createdAt',
    order = 'desc',
  ): Promise<PageResponse<OpeningListItem>> {
    return api.get<PageResponse<OpeningListItem>>(
      `/v1/public/openings?page=${page}&size=${size}&sort=${sort}&order=${order}`,
    );
  },

  async getPublicOpening(id: string): Promise<OpeningDetail> {
    return api.get<OpeningDetail>(`/v1/public/openings/${id}`);
  },

  async searchPublicOpenings(params: {
    q?: string;
    ecoCode?: string;
    moves?: string;
    page?: number;
    size?: number;
    sort?: string;
    order?: string;
  }): Promise<PageResponse<OpeningListItem>> {
    const searchParams = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== '') searchParams.append(key, String(value));
    });
    return api.get<PageResponse<OpeningListItem>>(
      `/v1/public/openings/search?${searchParams.toString()}`,
    );
  },
};

File: backend/src/main/java/com/chessot/service/PublicOpeningService.java
package com.chessot.service;

import com.chessot.dto.response.OpeningDetailResponse;
import com.chessot.dto.response.OpeningListItemResponse;
import com.chessot.dto.response.PageResponse;

import java.util.UUID;

public interface PublicOpeningService {

    PageResponse<OpeningListItemResponse> getPublicOpenings(
            int page, int size, String sort, String order);

    OpeningDetailResponse getPublicOpening(UUID id);

    PageResponse<OpeningListItemResponse> searchPublicOpenings(
            String query, String ecoCode, String moves,
            int page, int size, String sort, String order);
}
